[JENKINS-58362] Need to have only one copy of SystemCommandLanguage among detached plugins

jglick opened the pull request:
[JENKINS-58362](https://issues.jenkins-ci.org/browse/JENKINS-58362)

After https://github.com/jenkinsci/command-launcher-plugin/pull/4 and https://github.com/jenkinsci/jenkins/pull/4000/commits/b93ab72e6796e2bd611b14e86b262f897a01605f in #4000 we could have two copies of `SystemCommandLanguage` in the system, leading to weird situations. Fixes

```
mvn -f ../command-launcher-plugin surefire:test -Djenkins.version=2.184-SNAPSHOT -Dtest=CommandLauncher2Test\#requireApproval
```

since otherwise the test loads an old detached version of `command-launcher` (due to another bug: #4100) and we get

```
expected:<org.jenkinsci.plugins.scriptsecurity.scripts.languages.SystemCommandLanguage@…> but was:<org.jenkinsci.plugins.command_launcher.SystemCommandLanguage@…>
```

Yes this would have been caught some time ago by running `plugin-compat-tester` regularly.

Proposed changelog entry:
* Due to a mixture of old and new “detached” plugins, under certain circumstances it was possible to have two versions of an extension point used by the command launcher option for agents.

jglick committed:
Need to have only one copy of SystemCommandLanguage among detached plugins.

jglick commented on war/pom.xml:
```diff
@@ -405,7 +405,7 @@ THE SOFTWARE.
                   <groupId>org.jenkins-ci.plugins</groupId>
                   <artifactId>command-launcher</artifactId>
-                  <version>1.0</version>
+                  <version>1.2</version>
```
[Diff](https://github.com/jenkinsci/command-launcher-plugin/compare/command-launcher-1.0...command-launcher-1.2)

raul-arabaolaza reviewed (approved):
This fixes some broken internal PCT runs

jglick commented:
@olivergondza I am proposing this for backport to 2.176.x, maybe to .2 if it is not too late.

jglick committed:
Merge branch 'master' into SystemCommandLanguage

jglick committed:
Removing some cases from LoadDetachedPluginsTest which were sensitive to particular versions of detached plugins and thus hard to maintain.
Already had to do one such change in 84f939cebd1a0d03633e7b7c01e71ac387fb6463 and now upgradeFromJenkins2WithNewerDependency began failing after a war/pom.xml change.

jglick commented on test/src/test/java/jenkins/install/LoadDetachedPluginsTest.java:
```diff
@@ -87,62 +86,6 @@ public void upgradeFromJenkins2() {
-    @Issue("JENKINS-48604")
-    @Test
-    @LocalData
-    public void upgradeFromJenkins2WithNewerDependency() {
```
@dwnusbaum FYI I think these were your tests. Not sure how to rework them to be less brittle.

dwnusbaum replied:
Yeah, I don't have any good idea myself. The tests either need to use actual detached plugins or dynamically set up some simple, self-contained plugins as if they were detached plugins and use those in tests. The first option is brittle, and the second seems complicated.

jglick replied:
Right, the second option sort of presumes a test harness that would let you work with a specially-built WAR file. Beyond the power of `JenkinsRule`.

olivergondza commented:
@jglick, lets get this integrated into master, first. I do not see a reason this would not make it to .3, but you are right is is too late for .2.

oleg-nenashev commented:
@jglick  please write a changelog entry for the fix, see the PR template.
I will merge this change tomorrow if no negative feedback
